internal/status/status.go

```diff
@@ -0,0 +1,318 @@
+		return nonAvailableConditions
+	}
+	// get GC deployment conditions
+	nonAvailableConditions = s.helper.getDeploymentNotAvailableConditions(ctx, nfdInstance.Namespace, "nfd-gc")
+	if nonAvailableConditions != nil {
+		return nonAvailableConditions
+	}
+	// get topology, if needed
+	if nfdInstance.Spec.TopologyUpdater {
+		nonAvailableConditions := s.helper.getDaemonSetNotAvailableConditions(ctx, nfdInstance.Namespace, "nfd-topology-updater")
+		if nonAvailableConditions != nil {
+			return nonAvailableConditions
+		}
+	}
+
+	return getAvailableConditions()
+}
+
+func (s *status) AreConditionsEqual(prevConditions, newConditions []metav1.Condition) bool {
+	for _, newCondition := range newConditions {
+		oldCondition := meta.FindStatusCondition(prevConditions, newCondition.Type)
+		if oldCondition == nil {
+			return false
+		}
+		// Ignore timestamps
+		if oldCondition.Status != newCondition.Status ||
+			oldCondition.Reason != newCondition.Reason ||
+			oldCondition.Message != newCondition.Message {
+			return false
+		}
+	}
+	return true
+}
+
+//go:generate mockgen -source=status.go -package=status -destination=mock_status.go statusHelperAPI
+
+type statusHelperAPI interface {
+	getDaemonSetNotAvailableConditions(ctx context.Context, namespace, name string) []metav1.Condition
+	getDeploymentNotAvailableConditions(ctx context.Context, namespace, name string) []metav1.Condition
+}
+
+type statusHelper struct {
+	deploymentAPI deployment.DeploymentAPI
+	daemonsetAPI  daemonset.DaemonsetAPI
+}
+
+func newStatusHelperAPI(deploymentAPI deployment.DeploymentAPI, daemonsetAPI daemonset.DaemonsetAPI) statusHelperAPI {
+	return &statusHelper{
+		deploymentAPI: deploymentAPI,
+		daemonsetAPI:  daemonsetAPI,
+	}
+}
+
+func (sh *statusHelper) getDaemonSetNotAvailableConditions(ctx context.Context, dsNamespace, dsName string) []metav1.Condition {
+	failedToGetDSReason := conditionFailedGettingNFDWorkerDaemonSet
+	dsDegradedReason := conditionNFDWorkerDaemonSetDegraded
+	dsProgressingReason := conditionNFDWorkerDaemonSetProgressing
+	if dsName == "nfd-topology-updater" {
+		failedToGetDSReason = conditionFailedGettingNFDTopologyDaemonSet
+		dsDegradedReason = conditionNFDTopologyDaemonSetDegraded
+		dsProgressingReason = conditionNFDTopologyDaemonSetProgressing
+	}
+	ds, err := sh.daemonsetAPI.GetDaemonSet(ctx, dsNamespace, dsName)
+	if err != nil {
+		return getDegradedConditions(failedToGetDSReason, err.Error())
+	}
+	conditionsStatus, message := getDaemonSetConditions(ds)
+	if conditionsStatus == conditionStatusDegraded {
+		return getDegradedConditions(dsDegradedReason, message)
+	} else if conditionsStatus == conditionStatusProgressing {
+		return getProgressingConditions(dsProgressingReason, message)
+	}
+	return nil
+}
+
+func (sh *statusHelper) getDeploymentNotAvailableConditions(ctx context.Context, deploymentNamespace, deploymentName string) []metav1.Condition {
+	failedToGetDeploymentReason := conditionFailedGettingNFDMasterDeployment
```

Comment:
We use this for GC as well so we shouldn't unconditionally do `conditionFailedGettingNFDMasterDeployment`?

Reply:
The function checks what is the deployment (master or gc) and if needed changes the Reason. Please see line 181

Reply:
actually, my current code is sorta of ugly, i am changing it